cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -1194,3 +1197,30 @@ func (csr *ClusterStateRegistry) GetScaleUpFailures() map[string][]ScaleUpFailur
 	}
 	return result
 }
+
+func (csr *ClusterStateRegistry) fixupPlaceholderNodesForBackOff(instances map[string][]cloudprovider.Instance, currentTime time.Time) {
+	// Unfortunately the AWS cloud provider returns placeholder instances when an ASG's desired capacity is greater
+	// than the current. This is a problem because these nodes would still be considered by the logic that removes nodes
+	// failing to join the cluster, and this would intern scale down the backed-off ASGs that we want to keep scaled up.
+	// Unfortunately, these instances are also not marked in any way so it's impossible to distinguish them from normal
+	// ones (there's cloudprovider.Instance::Status but that doesn't seem to be used), and we would still like to
+	// terminate actual instances that fail to come up.
+
+	for _, nodeGroup := range csr.cloudProvider.NodeGroups() {
+		if csr.backoff.IsBackedOff(nodeGroup, csr.nodeInfosForGroups[nodeGroup.Id()], currentTime) {
+			for _, instance := range instances[nodeGroup.Id()] {
+				if unregisteredNodeEntry, ok := csr.unregisteredNodes[instance.Id]; ok {
+					// Figure out if it's a placeholder instance. Since there's no way to do this correctly, and I don't
+					// want to add another method to cloudprovider.CloudProvider, let's just hardcode the AWS logic.
+					if !strings.Contains(instance.Id, awsCloudProviderPlaceholderTag) {
```

Comment:
I think it would be easier to understand to flip this if statement since there is no else part, i.e.

```go
// If node is a placeholder, reset the unregistered node timeout so it's not deleted.
if strings.Contains(instance.Id, awsCloudProviderPlaceholderTag) {
  unregisteredNodeEntry.UnregisteredSince = currentTime
  csr.unregisteredNodes[instance.Id] = unregisteredNodeEntry
}
```

Reply:
Good point, fixed.